internal/storage/fs/cache.go

```diff
@@ -0,0 +1,198 @@
+// The supplied *reference* will NEVER be evicted.
+// Subsequent calls to Add with the same value for ref will update the fixed entries.
+// This should be called during initialization of a store.
+func (c *SnapshotCache[K]) AddFixed(ctx context.Context, ref string, k K, s *Snapshot) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.fixed[ref] = k
+	c.store[k] = s
+}
+
+// AddOrBuild adds the reference, key and value tuple.
+// If the reference r is already tracked in the fixed set, then it is updated there.
+// Otherwise, the entry is added to the LRU cache.
+func (c *SnapshotCache[K]) AddOrBuild(ctx context.Context, ref string, k K, build CacheBuildFunc[K]) (*Snapshot, error) {
+	s, ok, err := c.getByRefAndKey(ref, k)
+	if err != nil {
+		return s, err
+	}
+
+	// fast path: ref and key already exist and point to a valid snapshot
+	if ok {
+		return s, nil
+	}
+
+	// we build a new snapshot if getByRefAndKey failed to return one from
+	// the cache for the key k
+	if s == nil {
+		s, err = build(ctx, k)
+		if err != nil {
+			return s, err
+		}
+	}
+
+	// obtain a write lock to update all references to match the requested
+	// ref, key and snapshot
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	previous, ok := c.fixed[ref]
+	if ok {
+		// reference exists in the fixed set, so we updated it there
+		c.fixed[ref] = k
+	} else {
+		// otherwise, we store the reference in the extra LRU cache
+		// we peek at the existing value before replacing it so we
+		// can compare the previous value for deciding on eviction
+		previous, ok = c.extra.Peek(ref)
+		c.extra.Add(ref, k)
+	}
+
+	// update snapshot map using provided key
+	c.store[k] = s
+
+	// if the provided (new) target key (k) does not match the
+	// (now) previous key pointed at by ref then we attempt to
+	// evict the snapshot pointed at by the previous key
+	if ok && k != previous {
+		c.evict(ref, previous)
+	}
+
+	return s, nil
+}
+
+// Get attempts to resolve a snapshot for a given reference r.
+func (c *SnapshotCache[K]) Get(ref string) (s *Snapshot, ok bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	k, ok := c.fixed[ref]
+	if ok {
+		s, ok = c.store[k]
+		return
+	}
+
+	k, ok = c.extra.Get(ref)
+	if !ok {
+		return s, ok
+	}
+
+	s, ok = c.store[k]
+	return
+}
+
+func (c *SnapshotCache[K]) getByRefAndKey(ref string, k K) (s *Snapshot, ok bool, err error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	s, ok = c.store[k]
+	if !ok {
+		// return early as there is no k to snapshot mapping
+		// the snapshot is nil to signify it needs to be built
+		return
+	}
+
+	// check whether r is present in the fixed set
+	// if it is and it point to the provided k then
+	// we return true to signify everything is consistent
+	if fk, present := c.fixed[ref]; present {
+		ok = fk == k
+		return
+	}
+
+	// same as before except this time we check in the LRU
+	ek, present := c.extra.Get(ref)
+	// only return ok if both present in the LRU and existing key matches
+	ok = present && ek == k
+
+	return
+}
```

Comment:
Do we need `ok` or `err` to be returned from this function? I wonder if the caller could just rely on if `s` is nil or not?

Reply:
Actually err can be dropped! I rewrote this a few times, that is a vestigial limb from an old iteration.

`ok` on the other hand signifies something important.
It signifies both that `k` is in the cache (so `s` is not nil) **and** that ref ponts at `k` either from the fixed map or from the LRU.

If `ok` then we can return early and we only obtained a read lock.
However, if `!ok` then we potentially need to build `s` and/or update what `ref` points to.